.github/workflows/pr.yml

```
@@ -1,21 +1,27 @@
+---
 name: PR testing
 
-on: [push, pull_request]
+# run on each PR and on merges to main
+on:
+  pull_request: {}
+  push:
+    branches:
+      - main
 
 jobs:
   test:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-24.04
     strategy:
       matrix:
-        java: [ '11, '17' ]
+        java: [ '11', '17', '21' ]
```

I always prefer multi line statements since the diffs are easier to read.
```suggestion
        java:
          - '11'
          - '17'
          - '21'
```